### ISSUE TITLE
Occasional crash in duk_heap_mark_and_sweep

### ISSUE THREAD

**Issue by beevik** [svaarala/duktape#1375]
**duktape version**: 2.0.0

I am ocassionally seeing a crash in `duk_heap_mark_and_sweep`. The crash is triggered by dereferencing a null pointer.

Here is the code segment in question:

```c
#if defined(DUK_USE_FINALIZER_SUPPORT)
	duk__mark_finalizable(heap);              /* mark finalizable as reachability roots */
	duk__mark_finalize_list(heap);             // CRASH HAPPENING HERE
#endif
	duk__mark_temproots_by_heap_scan(heap);   /* temproots */
```

Here is the disassembly of the code fragment above.  The `duk__mark_finalize_list` function has been inlined, and the crash appears to be happening as the `heap->finalize_list` is being traversed.  The location of the crash is marked with "CRASH HAPPENING HERE".

```
#if defined(DUK_USE_FINALIZER_SUPPORT)
	duk__mark_finalizable(heap);              /* mark finalizable as reachability roots */
00007FFF033134EE  mov         rcx,rbx  
00007FFF033134F1  call        duk__mark_finalizable (07FFF03312BC0h)  
	duk__mark_finalize_list(heap);            /* mark finalizer work list as reachability roots */
00007FFF033134F6  mov         rdi,qword ptr [rbx+58h]  
00007FFF033134FA  test        rdi,rdi  
00007FFF033134FD  je          duk_heap_mark_and_sweep+104h (07FFF03313544h)  
00007FFF033134FF  mov         eax,dword ptr [rdi]         ;  ***CRASH HAPPENING HERE***
00007FFF03313501  test        al,4  
00007FFF03313503  jne         duk_heap_mark_and_sweep+0FBh (07FFF0331353Bh)  
00007FFF03313505  or          eax,4  
	duk__mark_finalize_list(heap);            /* mark finalizer work list as reachability roots */
00007FFF03313508  mov         dword ptr [rdi],eax  
00007FFF0331350A  mov         eax,dword ptr [rbx+4Ch]  
00007FFF0331350D  cmp         eax,100h  
00007FFF03313512  jl          duk_heap_mark_and_sweep+0DCh (07FFF0331351Ch)  
00007FFF03313514  or          dword ptr [rbx],2  
00007FFF03313517  or          dword ptr [rdi],8  
00007FFF0331351A  jmp         duk_heap_mark_and_sweep+0FBh (07FFF0331353Bh)  
00007FFF0331351C  inc         eax  
00007FFF0331351E  mov         dword ptr [rbx+4Ch],eax  
00007FFF03313521  mov         eax,dword ptr [rdi]  
00007FFF03313523  and         eax,3  
00007FFF03313526  je          duk_heap_mark_and_sweep+0F8h (07FFF03313538h)  
00007FFF03313528  cmp         eax,1  
00007FFF0331352B  jne         duk_heap_mark_and_sweep+0F8h (07FFF03313538h)  
00007FFF0331352D  mov         rdx,rdi  
00007FFF03313530  mov         rcx,rbx  
00007FFF03313533  call        duk__mark_hobject (07FFF03312D00h)  
00007FFF03313538  dec         dword ptr [rbx+4Ch]  
00007FFF0331353B  mov         rdi,qword ptr [rdi+10h]  
00007FFF0331353F  test        rdi,rdi  
00007FFF03313542  jne         duk_heap_mark_and_sweep+0BFh (07FFF033134FFh)  
#endif
	duk__mark_temproots_by_heap_scan(heap);   /* temproots */
```

I am not able to come up with a reliable reproduction of this crash.

**Comment by svaarala** [comment]
There are a few related fixes (e.g. #1347) in 2.0.1 and in v2.0-maintenance branch -- could you try to reproduce with e.g. v2.0-maintenance branch?

**Comment by beevik** [comment]
will try tomorrow.

**Comment by svaarala** [comment]
Ok, thanks @beevik. Also a C stack trace would be helpful if you can produce one easily.

**Comment by beevik** [comment]
Here is the stack:

```
Engine!duk_heap_mark_and_sweep
Engine!duk_refzero_free_pending
Engine!duk_set_top
Engine!duk__reconfig_valstack_ecma_return
Engine!duk__handle_return
Engine!duk__js_execute_bytecode_inner
Engine!duk_js_execute_bytecode
Engine!duk__handle_call_inner
Engine!duk_handle_call_protected
```

**Comment by beevik** [comment]
And here's the stack from a separate crash in the same location:

```
Engine!duk_heap_mark_and_sweep
Engine!duk_heap_mem_alloc
Engine!duk_hobject_alloc
Engine!duk_push_object
Engine!duk_new
Engine!duk__js_execute_bytecode_inner
Engine!duk_js_execute_bytecode
Engine!duk__handle_call_inner
Engine!duk_handle_call_protected
```

**Comment by svaarala** [comment]
The former is quite possibly caused by #1347 (voluntary mark-and-sweep triggered by refzero processing). The latter doesn't look like it directly, but it could still have the same root cause: #1347 could cause the heap to be in inconsistent state and that might only manifest later.

**Comment by beevik** [comment]
Ok I've upgrade to v2.0.1.  I'll close this issue. If we see the crash again, I'll reopen.

**Comment by fatcerberus** [comment]
> upgraded to v2.0.1

And you're sure that fixed the crashes for you?  The bug in #1347 is present in v2.0.1 as well.

**Comment by svaarala** [comment]
If it helps any, I can push out 2.0.2 with the fix included. While I'd like to resolve a few more issues before the next maintenance release, there are enough fixes in 2.0.2 and 1.7.0 to justify a release right away too.

**Comment by beevik** [comment]
No hurry, I can wait.

**Comment by beevik** [comment]
Reopening after getting a similar crash with the cherry-picked fixes.

Here is the fork of the repo we're using: https://github.com/arenanet/duktape/commits/v2.0.1-arenanet

As you can see I cherry-picked the recursive mark-and-sweep fixes.

The stack we got on the most recent crash is:

```
duk__mark_heaphdr(duk_heap * heap=0x00000282dc89c678, duk_heaphdr * h=0xcccccccc00000000) Line 165               C
duk__mark_finalize_list(duk_heap * heap=0x00000282dc89c678) Line 353  C
duk_heap_mark_and_sweep(duk_heap * heap=0x00000282dc89c678, unsigned int flags=0) Line 1282       C
duk_refzero_free_pending(duk_hthread * thr=0x00000282e2406568) Line 444         C
duk_set_top(duk_hthread * ctx=0x00000282e2406568, int idx=10) Line 475               C
duk__handle_call_inner(duk_hthread * thr=0x00000282e2406568, int num_stack_args=2, unsigned int call_flags=0, int idx_func=9) Line 1683           C
duk_handle_call_unprotected(duk_hthread * thr=0x00000282e2406568, int num_stack_args=2, unsigned int call_flags=0) Line 1154          C
duk__js_execute_bytecode_inner(duk_hthread * entry_thread=0x00000282e2406568, unsigned __int64 entry_callstack_top=1) Line 4544              C
duk_js_execute_bytecode(duk_hthread * exec_thr=0x00000282e2406568) Line 2353            C
duk__handle_call_inner(duk_hthread * thr=0x00000282e2406568, int num_stack_args=0, unsigned int call_flags=0, int idx_func=0) Line 1569           C
duk_handle_call_protected(duk_hthread * thr=0x00000282e2406568, int num_stack_args=0, unsigned int call_flags=0) Line 1054          C
duk_pcall_method(duk_hthread * ctx=0x00000282e2406568, int nargs=0) Line 159 C
```

**Comment by beevik** [comment]
By the way, that stack shows the parameter values accurately, so it appears the `h` parameter of `duk__mark_heaphdr` has a busted pointer (`0xcccccccc00000000`).

I have a full memory dump of the crash, so I'll do some spelunking to see if there's any more info I can draw out of it.

**Comment by svaarala** [comment]
Is it possible to run/repeat with assertions enabled (if you already didn't do that)?

**Comment by svaarala** [comment]
Just to check: do you use finalizers, and if so, are there any suspend/resume calls in finalizers?

**Comment by beevik** [comment]
We run with finalizers on some objects.  There is never a suspend or resume call within a finalizer.  suspend and resume are always done outside the calling of methods.

Unfortunately, I cannot replay the crash.

**Comment by svaarala** [comment]
Re: suspend/resume, if you switch native threads with the Duktape call stack empty, there's no need to suspend/resume. The calls are only needed if you suspend from an active call stack; otherwise it's sufficient just to ensure only one native thread calls into Duktape at a time (but I might have misunderstood you). I asked about finalizers specifically because suspending from a finalizer doesn't work at present, but that's not an issue here.

Without a repro case it's difficult to say what the ultimate cause could be, but the stack trace indicates that:
- Refzero (freeing of objects whose refcount is now zero) is in progress
- Mark and sweep is triggered during that process (code line indicates "voluntary mark-and-sweep")

There was a bug there causing a mark-and-sweep re-entry, but the cherry-picked fixes should have taken care of that, so I'm not sure what the remaining cause might be.

There are some unfixed side effect issues in #1311 however, so maybe one of them is related to your case. My next priority is actually to work on fixing the issues in #1311, and to improve torture test coverage so that it would have caught the issues. The difficult thing with side effect related bugs is that the trigger conditions are relatively rare, so that ordinary test case runs (even if quite extensive) won't necessarily ever cause those conditions, so forced torture tests are needed.

**Comment by svaarala** [comment]
If the root cause is voluntary GC related, one workaround you may be able to use is disable `DUK_USE_VOLUNTARY_GC`. That means GC only runs when out-of-memory or explicitly requested. As a side effect objects participating in reference loops (like function instances) won't be freed until GC runs, so that either needs to be OK (sometimes it is, sometimes not, depends on your code base), or you need to call duk_gc() from time to time (say once a minute).

I'm not suggesting this as an actual solution - but may be worth trying if the issue is acute.

**Comment by beevik** [comment]
Re: switch/resume.  I have multiple "duktape threads", each of which runs inside its own native coroutine. Before switching processing to another coroutine, I have to call `duk_suspend` on the old coroutine; then I call `duk_resume` on the newly activated coroutine.  If I don't do this, bad things happen.  So it appears that even though this is all happening on one native thread, the calls to `duk_suspend` and `duk_resume` are necessary.

It's possible there's a memory stomp issue in my own code that is causing this problem. But the crash always seems to manifest in the same way: by blowing up while accessing `heap->finalize_list`.  This is what leads me to think it more likely that there's some bug in the GC.

I'll continue trying to get a repro of this bug. My system is capable of playbacks, so I'll make sure they're always being recorded.  And if I'm lucky I'll get a recording of this crash.

**Comment by svaarala** [comment]
> Re: switch/resume. I have multiple "duktape threads", each of which runs inside its own native coroutine. Before switching processing to another coroutine, I have to call duk_suspend on the old coroutine; then I call duk_resume on the newly activated coroutine. If I don't do this, bad things happen. So it appears that even though this is all happening on one native thread, the calls to duk_suspend and duk_resume are necessary.

The critical thing when not using suspend/resume is that only a single native thread calls into Duktape. Before any other native thread can call into Duktape, the current one must return from whatever calls are in progress, so that (1) there are no active call stack entries (= no functions running in the script side), (2) no Duktape API calls are in progress. I understood this to be the case when you said "suspend and resume are always done outside the calling of methods" -- but I'm not sure if that's what you meant.

Suspend/resume is needed when the native thread calling into Duktape changes but there are still active calls in the Duktape call stack of the current (Duktape) thread.

**Comment by svaarala** [comment]
Clarification to above: there shouldn't be any active call stack entries, but also no active "safe calls" (which don't involve a call stack entry but do involve call state).

**Comment by beevik** [comment]
> Suspend/resume is needed when the native thread calling into Duktape changes but there are still active calls in the Duktape call stack of the current (Duktape) thread.

Ok it makes sense now. Right now, when one of my duktape scripts "yields" its coroutine, there are definitely calls active on the duktape call stack.  This is when I call `duk_suspend`, followed by `duk_resume` in another coroutine to resume execution of its own previously suspended duktape call stack.

**Comment by beevik** [comment]
More info.  So I turned on assertions, and I'm finding that I sometimes hit the assertion on the marked line in `duk__validate_push_heapptr`:

```
	for (curr = thr->heap->finalize_list;
	     curr != NULL;
	     curr = DUK_HEAPHDR_GET_NEXT(thr->heap, curr)) {
		if (curr == h) {
			if (DUK_HEAPHDR_HAS_FINALIZED((duk_heaphdr *) h)) {
				/* Object is currently being finalized. */
				DUK_ASSERT(found == 0);  /* Would indicate corrupted lists. */
				found = 1;
			} else {
				DUK_ASSERT(0);   /*** HITTING THIS ASSERTION ***/
			}
		}
	}
```

This is happening in a call to `duk_push_heapptr`.  Any idea why this might be happening?

**Comment by svaarala** [comment]
The argument to `duk_push_heapptr()` should have a guarantee from the application that it has been reachable at all times between a `duk_get_heapptr()` and `duk_push_heapptr()`. If the argument is in the finalize_list, it has apparently become unreachable between the two calls which the assert tests for.

That's the public API guarantee at present.

The assert does make an exception for the case where a finalizer call uses `duk_push_heapptr()` for the finalizer argument itself. In this case the argument has become unreachable between `duk_get_heapptr()` and `duk_push_heapptr()`, but nothing unsafe happens so it's not rejected.

The `DUK_ASSERT(0)` you're hitting is the case where the argument has become unreachable (because it's in finalize_list) and it's not the object currently being finalized.

**Comment by beevik** [comment]
Is it possible for an object to be on the finalizer list without me having received the finalize callback?  This object that's showing up in the assertion has a finalizer callback, which was set with a call to `duk_set_finalizer`.  The only reason I would even attempt to access its heapptr is if I had not yet been told it was finalized.

**Comment by svaarala** [comment]
It's possible: objects are placed in the finalize_list when they become unreachable but haven't had their finalizers called yet. There's no easy way for an application to notice when an object has become unreachable but hasn't yet been finalized. It may take some time for an object to be finalized, especially if mark-and-sweep does it (for objects in reference cycles).

Ideally pushing an unreachable object that hasn't yet been finalized would simply rescue the object in question. But this doesn't work at present: the object would get back into play but wouldn't be removed from finalize_list, breaking various internal book-keeping.

**Comment by beevik** [comment]
Is there a way to query whether an object is on the finalizer list?

**Comment by svaarala** [comment]
No, there's no current way of knowing that an object has become unreachable but hasn't yet been finalized. The API contract is that an application is required to ensure reachability at all times between `duk_get_heapptr()` and `duk_push_heapptr()` so that the object cannot be in the finalize_list.

In other words, if the object is on the finalize_list the application has already broken that API contract.

There's an exception in the assert for pushing the object being finalized - while this is a break in the API contract, it does work at present because the book-keeping comes out all right for the object being finalized. In particular, the object gets queued back into the main heap object list normally once the finalizer returns.

**Comment by svaarala** [comment]
As a technical note, an object may be finalized by the refcounter (refzero_list) or mark-and-sweep (finalize_list). The issues are similar: if duk_push_heapptr() pushes a pointer in either list the internal book-keeping won't come out right (except for the object being finalized right now).

**Comment by beevik** [comment]
How would you recommend creating a C-backed object whose lifetime is bounded by its script-side counterpart?  Right now, I have implemented a system where I call `duk_get_heapptr` after calling `duk_new` to create an object.  I then freely use this heapptr until I receive a finalize callback for that object. If I cannot know that the heapptr has been placed on a finalize list, I'm not sure how to know when to destroy the C-backed object.

**Comment by svaarala** [comment]
If the use case calls for true weak references, i.e. borrowed references obtained using duk_get_heapptr() without a corresponding strong reference to guarantee reachability, I don't think that's unfortunately possible at present. The main problem is the intermediate time between an object becoming unreachable without yet being finalized.

But the main question really is how (and why) your code is using duk_push_heapptr().

**Comment by fatcerberus** [comment]
> How would you recommend creating a C-backed object whose lifetime is bounded by its script-side counterpart?

You do that by allocating the C-backed object in the constructor, and freeing it in the finalizer.  At that point you're at the mercy of the garbage collector; trying to do it from the other direction won't work.  Thankfully Duktape is usually pretty quick about cleaning things up.

For example, in miniSphere's `Texture` constructor:
https://github.com/fatcerberus/minisphere/blob/master/src/minisphere/pegasus.c#L2128

The `image_t*` pointer is associated with the script-side object.  In the `Texture` finalizer:
https://github.com/fatcerberus/minisphere/blob/master/src/minisphere/pegasus.c#L2137-L2138

The reference is released.

**Comment by beevik** [comment]
> You do that by allocating the C-backed object in the constructor, and freeing it in the finalizer. At that point you're at the mercy of the garbage collector; trying to do it from the other direction won't work. Thankfully Duktape is usually pretty quick about cleaning things up.

That's exactly what I'm doing. Unfortunately, it is never safe to use the script object's heapptr in this scenario, because who knows when the script object entered the finalize list.

**Comment by svaarala** [comment]
@fatcerberus I think the issue is not so much freeing the native resources, but about how to hold a true weak reference (borrowed pointer without a known strong reference backing it up) in C code.

**Comment by beevik** [comment]
>  I think the issue is not so much freeing the native resources, but about how to hold a true weak reference (borrowed pointer without a known strong reference backing it up) in C code.

It seems that the issue is not so much that it's a "weak reference" to a script object, but that the object's heapptr becomes invalid before the finalize callback is made.   If it were safe to use the heapptr until the finalize callback, there'd be no issue.  But I guess that once the object has become "unreachable", its heapptr is invalid, even if the finalize callback hasn't happened yet.

**Comment by fatcerberus** [comment]
@beevik For what purpose are you holding onto the heapptr, exactly?  I'm trying to understand the use case.

**Comment by fatcerberus** [comment]
> I want to be able to pass references to the script object from C into my script. To do this, I need the heapptr. If I were to use something like a stash property instead of a heapptr, then duktape would no longer control the object's lifetime; my C code would.

I had a feeling this was the case.  Essentially what you want is a guarantee that the JS object should last at least as long as the C object.  Unfortunately that's not currently possible without using the stash; which as you said conflicts with your first point (letting Duktape control the lifetime).  I don't even think weak references would solve that since by definition a weak reference can't prevent garbage collection.

**Comment by svaarala** [comment]
@beevik I'm using "weak" reference to mean a pointer that is not backed by a strong reference visible to Duktape GC. The API assumes the object is reachable all the way between duk_get_heapptr() and duk_push_heapptr(). This in essence means that the reference must be backed by an actual, GC visible (strong) reference between the two calls.

To allow actual weak references to be held by C code, the two main possibilities would be:
- Allow duk_push_heapptr() for unreachable objects before they are finalized, and automatically cancelling the finalization in this case.
- Add a separate notification indicating than an object became unreachable (queued to refzero_list or detected as unreachable by mark-and-sweep). C code would then need to be able to locate its own state based on the pointer alone, which is probably not very convenient.

The former would be most convenient and intuitive for C code. There are some internal book-keeping concerns in the change, e.g. the linked lists are one way (at least with refcounting disabled) so removing an element from the middle requires a scan from the beginning. Finalize_list and refzero_list are typically not very long, so this might be OK.

However, weak references may also be useful with objects that don't have a finalizer, and adding a finalizer just to get the "unreachable" notification may be overkill for some use cases. So maybe both mechanisms might be needed.

**Comment by beevik** [comment]
> don't even think weak references would solve that since by definition a weak reference can't prevent garbage collection.

A weak reference would at least let me check whether it's safe to call `duk_push_heapptr` before doing so.

**Comment by beevik** [comment]
>There are some internal book-keeping concerns in the change, e.g. the linked lists are one way (at least with refcounting disabled) so removing an element from the middle requires a scan from the beginning. Finalize_list and refzero_list are typically not very long, so this might be OK.

I thought it was a doubly-linked list.  Isn't that what DUK_USE_DOUBLE_LINKED_HEAP is for?  Or were you referring to some other list?

**Comment by svaarala** [comment]
> I could not find any description of this issue with heapptr lifetimes and finalizers in the documentation. It might be worth adding something about it. I was operating under the assumption that a heapptr would be valid until the finalizer was called, but that's clearly not the case.

The API documentation for duk_push_heapptr() tries to be explicit about the reachability requirement:

> The argument ptr must be obtained using duk_get_heapptr() (or its variants) and must have been reachable for Duktape garbage collection up to this call.

I can appreciate why the implications are not very clear from that, opened #1392 to clarify the finalizer issue in particular in the API documentation.

**Comment by fatcerberus** [comment]
Yeah, I can see the room for confusion there.  It'd be easy for someone to assume that, because the finalizer is passed a reference to the object being finalized, and can freely access its properties up until the finalizer returns--and even rescue it--it's therefore still reachable "by definition".

**Comment by svaarala** [comment]
> I thought it was a doubly-linked list. Isn't that what DUK_USE_DOUBLE_LINKED_HEAP is for? Or were you referring to some other list?

When refcounting is disabled, the heap is single linked and that's one supported configuration. With refcounting enabled there are prev/next pointers, but it's still possible to use a single linked list for e.g. finalize_list or refzero_list to improve list handling slightly.

Right now, in master, finalizer_list and refzero_list are double linked with refcounting enabled, but I have a work branch where refzero_list would become single linked to improve GC performance slightly.

**Comment by svaarala** [comment]
> [...] and can freely access its properties up until the finalizer returns--and even rescue it--it's therefore still reachable "by definition".

Right, it gets a bit technical - reachability by definition means that there's a path from some reachability root to the object. Quite clearly the most intuitive solution would be to eliminate the "gap" between an object becoming unreachable and being finalized.

(And in fact heap->finalize_list *is* a reachability root for mark-and-sweep so that otherwise unreachable objects and whatever they reference won't be freed until the finalizer has been executed.)

**Comment by svaarala** [comment]
> Does this mean the assertion is just an extra safety check but that referencing the heapptr of an object on the finalize list should still be expected to work?

There was a previous issue where a heapptr of an unreachable object was pushed in the same way, and it did cause similar issues (I added the detailed duk_push_heapptr() asserts at that time). So quite possibly this is/was causing the issues you saw. I'd need to read the code carefully to figure out what specifically goes wrong when the assertions are violated, but in that previous case the internal linked lists became corrupted (spliced) which causes odd issues and memory unsafe behavior after some delay (making diagnosis difficult).

At the very least the finalizer would execute even after an object had been pushed with duk_push_heapptr() which may cause issues in the application side, depending on how the finalizer was written.

**Comment by beevik** [comment]
Ok I'll stop bugging you today.  Thanks again for the help.

**Comment by svaarala** [comment]
I'll give some thought tomorrow to maybe relaxing that finalize_list limitation for the 2.1.0 release.

Maybe it would be enough to detect the situation, flag the object as "autorescued by duk_push_heapptr()", and when it's time to run the object's finalizer, check that flag and skip the finalizer if the flag is set. It's conceptually simple stuff but GC/finalizer code always needs careful thought.

**Comment by fatcerberus** [comment]
Wouldn't it be enough just to check, when the object is pushed, whether it's in the finalize_list and remove it if so?  Seems less error-prone than a flag.

**Comment by svaarala** [comment]
@fatcerberus The flag would avoid the dequeueing which would avoid walking the single linked list from the beginning. The control path the object takes would also be almost identical to a finalizer rescued object so there would be less need for test coverage.

**Comment by svaarala** [comment]
One thought that came to mind re: allowing duk_push_heapptr() for a finalize_list object:

Let's assume that the duk_push_heapptr() operation would automatically rescue the object without the finalizer being invoked. This would make the pushing safe -- but it would also mean that the C side pointer wouldn't be truly weak because using the pointer would extend the object's lifetime.

For mark-and-sweep the impact would be quite noticeable: the object would need to go through a few cycles for its finalizer to be executed (assuming it wasn't "auto rescued" by duk_push_heapptr() again).

Usually finalize_list is processed immediately after mark-and-sweep (this is actually also why @beevik's duk_gc() usually avoids the asserts) but not always. Also, one finalizer may have an effect on the other objects pending finalization.

So, this tweak wouldn't actually be a 100% match to allowing true weak pointers from C code.

It might seem easy to just check if the duk_push_heapptr()'d value was only temporarily needed (say, to log a property value) so that one could execute the finalizer without postponing it and thus affecting the object lifetime. But the only realistic way to check the safety of this is mark-and-sweep itself.

**Comment by fatcerberus** [comment]
In theory, allowing this should be safe regardless of when finalization happens, as long as the application stops using the pointer after the finalizer is called, right?  So I don't see why Duktape would have to delay finalization further?

**Comment by svaarala** [comment]
> So I don't see why Duktape would have to delay finalization further?

But wouldn't you find it odd if the finalizer is called for an object which was made (permanently) reachable by the application calling duk_push_heapptr()?

**Comment by fatcerberus** [comment]
> But wouldn't you find it odd if the finalizer is called for an object which was made (permanently) reachable by the application calling duk_push_heapptr()?

Indeed, but I thought that was accounted for by the hypothetical "autorescue" flag.  Your wording above seemed to imply that finalization would be delayed even for cases where no rescuing happened.  If that's not what you meant, then disregard. :)

**Comment by svaarala** [comment]
> Your wording above seemed to imply that finalization would be delayed even for cases where no rescuing happened.

It potentially would -- if the application calls duk_push_heapptr() before finalization takes place, the only reliable way of knowing whether the object is now reachable or not is to postpone the finalization and run another mark-and-sweep pass.

**Comment by fatcerberus** [comment]
Ah, I see.  So Duktape would need to run mark-and-sweep *again* to know if it's still safe to call the finalizer.  I understand the issue now.

**Comment by svaarala** [comment]
In some very simple cases like: `duk_push_heapptr(); duk_pop()` it might be possible to infer that the object's net reachability status hadn't changed. But in practical terms that's quite difficult to do reliably, especially because refcounts don't really help: objects in finalize_list are by design unreachable but with refcounts > 0 because they participate in reference loops.

**Comment by svaarala** [comment]
Re: finalizer delays, the common case is that there's no real delay:
- For refzero handling finalizers are executed inline.
- For mark-and-sweep finalizers are executed after the main mark-and-sweep pass.

The visible delays are caused by:
- There being multiple objects pending finalization (happens in both refcounting and mark-and-sweep). In this case one finalizer may reference another object still pending finalization.
- Mark-and-sweep being caused by a call site which prevents finalizer side effects so that the objects remain in finalize_list until the next mark-and-sweep. This introduces a relatively long delay.

The delays can't be fully eliminated (especially the first category - multiple objects may be pending finalization by design) but the latter mark-and-sweep related delay could be reduced considerably by running pending finalizers explicitly when the call site preventing finalizer execution finishes its critical segment (similarly to the forced refzero check in call sites using NORZ macros).

But this would just make the window for the issue smaller, so while this would be a useful improvement for snappier finalization, it's unrelated to the main safety issue in duk_push_heapptr().

**Comment by harold-b** [comment]
This all reminds me of #914 :)

**Comment by svaarala** [comment]
Being on finalize_list means that the object was unreachable at one point in time. It will remain unreachable until the finalizer is called because nothing can legitimately push the object back: if anyone had an actual, non-borrowed reference, the object wouldn't be on the finalize_list to begin with. So the object can only become reachable by using `duk_push_heapptr()` with a borrowed pointer which isn't backed up by a strong reference.

I'm working right now on fixing some side effect issues related to the finalizers, and as a by-product of that work all finalization (both refcount and mark-and-sweep) will go through finalize_list. It would then be relatively easy to ensure every object on finalize_list had a flag indicating they're on that list, and because the link is double linked, remove the object from the list prematurely.

This should allow an actual solution to be implemented, i.e. `duk_push_heapptr()` would transparently rescue an object.

It would also be easy to add an API call to check if an object is on finalize_list like you suggest -- but while that's relatively easy to implement, it would be an awkward API promise going forward when the internal implementation changes.

**Comment by beevik** [comment]
Got it, thanks for addressing this.  I was actually looking into what it would take to add a new flag to mark a heap object as "on the finalize list" and discovered that all 32 flags of the heaphdr are already in use!  heh.

**Comment by svaarala** [comment]
Yeah, the flags are pretty tightly used now. Some of the flags could be removed and be replaced with an internal class test, but that should only be done for performance insensitive paths because a mask+compare is worse than a bit or bit mask test. For example, I'm freeing up `DUK_HOBJECT_FLAG_THREAD` because it's only checked for in just a few places and a class number based test is good enough for them.

**Comment by svaarala** [comment]
@beevik As far as I can tell this has now been addressed in master -- if not, please reopen.

**Comment by beevik** [comment]
thanks!  Will let you know once 2.1 is released and we do our next duktape update.

**Comment by beevik** [comment]
I thought I should provide an update.

I've switched to 2.1 and my problems with duk_push_heapptr are no more.  The pointer rescue code is working.  Thanks!

**Comment by svaarala** [comment]
Thanks, that's nice to hear!

**Comment by mpietras** [comment]
For what it's worth, I just had a crash in 2.0.2 with stack like this:

h is invalid (0xb5ee2880380a80e9) at:
if (DUK_HEAPHDR_HAS_REACHABLE(h)) {
duk__mark_heaphdr(duk_heap * heap, duk_heaphdr * h) Line 42014
duk__mark_hobject(duk_heap * heap, duk_hobject * h) Line 41918 (i==3 so it's a few iterations in)
duk__mark_heaphdr(duk_heap * heap, duk_heaphdr * h) Line 42045 (not sure what's interesting here... heap->mark_and_sweep_recursion_depth==3, trigger_counter==6816, base_flags==0)
duk__mark_hobject(duk_heap * heap, duk_hobject * h) Line 41912 (i==50)
duk__mark_heaphdr(duk_heap * heap, duk_heaphdr * h) Line 42045
duk__mark_hobject(duk_heap * heap, duk_hobject * h) Line 41995 (i==0 bottom of hobj_is_thread)
duk__mark_roots_heap(duk_heap * heap) Line 42067
duk_heap_mark_and_sweep(duk_heap * heap, unsigned int flags) Line 43098

I'm reviewing duktape fixes to see if upgrade might have an impact but if something jumps out at you please let me know, thanks!

**Comment by svaarala** [comment]
@mpietras The major side effect fixes were implemented in 2.1.0 and were not backported to 2.0.x (the diff would have been large, and it would then be a bit questionable whether it'd be a "patch" release).

**Comment by mpietras** [comment]
Thanks, understood.  I was just looking for an opinion if this crash stack matched any of the characteristics of the issues fixed in 2.1.0... i.e. if I would expect that upgrading from 2.0.2 to 2.1.1 would possibly resolve crashes of this nature...

**Comment by svaarala** [comment]
The triggers for the issues before 2.1.0 were relatively complicated but they would typically require finalizer execution, possibly side effects within finalizers, and were more likely when objects were rescued from finalizers. But it's difficult to say if your case would match these without knowing more of the situation.
